chore(svelte): Hide cody nav entries and upsell banner when cody isn't enabled

fkling opened the pull request:
Contributes to srch-529

This commit refactors the main navigation and search home page code to make it more configurable. In particular we now only show navigation entries for features that are enabled (as determined by `window.context`) and only show the cody upsell banner when cody is enabled.

I extended the dev HTML template and .env files to support this.

| Command | Screen |
|--------|--------|
| `pnpm dev` | ![2024-06-25_10-57](https://github.com/sourcegraph/sourcegraph/assets/179026/7b14b8d0-617a-4397-89cd-e505603ef258) |
| `pnpm dev:dotcom` | ![2024-06-25_10-56_1](https://github.com/sourcegraph/sourcegraph/assets/179026/506fa5ce-577c-4aa9-8d99-4ae1094a37b9) |
| `PUBLIC_CODY_ENABLED= pnpm dev:dotcom` | ![2024-06-25_10-56](https://github.com/sourcegraph/sourcegraph/assets/179026/9bd914d1-3588-40d9-8938-56eabd7dedb9) | 

## Test plan

Manual testing.

fkling commented on client/web-sveltekit/src/routes/search/DotcomFooterLinks.svelte:
```diff
@@ -31,7 +29,9 @@
 
     function handleLinkClick(telemetryType: number): void {
-        TELEMETRY_RECORDER.recordEvent('home.footer.CTA', 'click', { metadata: { type: telemetryType } })
+        import('$lib/telemetry').then(({ TELEMETRY_RECORDER }) => {
+            TELEMETRY_RECORDER.recordEvent('home.footer.CTA', 'click', { metadata: { type: telemetryType } })
+        })
```
I got an error about `location` not being defined when `$lib/telemetry` is imported directly. That seems to be because `+page.ts` imports this component directly and seems to be executed in node somehow 🤷🏻

camdencheek replied:
I'm confused

fkling replied:
The tl;dr is that it's difficult to work with code that assumes to run in a browse environment *during initialization* (e.g. loading a module). That's why importing stuff from `@sourcegraph/wildcard` often results in errors because it accesses `window` during module initialization.
That's a problem with SvelteKit, and server side rendering in general. Even though I disabled server side rendering, some modules (such as the `+page|layout.ts` files) are still loaded in node during build time, which causes the same problem.

camdencheek replied:
Ah, gotcha. Would it be better to add the telemetry recorder as context on the root layout then?

fkling replied:
I'll have to think about this a bit, but that seems like a good idea to me.

fkling commented on client/web-sveltekit/src/routes/search/+page.svelte:
```diff
@@ -41,5 +41,12 @@
+            {/await}
+        {/if}
+        <svelte:component this={data.footer} />
+    </SearchHome>
```
This takes out all the cody and dotcom specific logic from `SearchHome`.

fkling commented on client/web-sveltekit/src/routes/search/+page.svelte:
```diff
@@ -41,5 +41,12 @@
     />
 {:else}
-    <SearchHome {queryState} codyHref={data.codyHref} showDotcomFooterLinks={data.showDotcomFooterLinks} />
+    <SearchHome {queryState}>
+        {#if data.isCodyEnabled}
+            {#await import('./cody-upsell/CodyUpsellBanner.svelte') then CodyUpsell}
+                <svelte:component this={CodyUpsell.default} codyHref={data.codyHref} />
+            {/await}
+        {/if}
+        <svelte:component this={data.footer} />
+    </SearchHome>
```
This takes out all the cody and dotcom specific logic from `SearchHome`.

camdencheek replied:
Ooh, I like that. Thanks for cleaning up my mess 🙂 

Could you elaborate a bit on the lazy import? Is this so we can reduce initial bundle size?

camdencheek commented on client/web-sveltekit/.env.dotcom:
```diff
@@ -1 +1,4 @@
 PUBLIC_DOTCOM=true
+PUBLIC_CODY_ENABLED=true
+PUBLIC_BATCH_CHANGES_ENABLED=
+PUBLIC_CODE_INSIGHTS_ENABLED=
```
Oh, this is nice!

camdencheek commented on client/web-sveltekit/src/routes/navigation.ts:
```diff
@@ -1,10 +1,27 @@
- * The main navigation of the application.
- */
-export const mainNavigation: (NavigationMenu | NavigationEntry)[] = [
+export const enum Mode {
+    ENTERPRISE = 1 << 0,
+    DOTCOM = 1 << 1,
+    CODY_ENABLED = 1 << 2,
+    BATCH_CHANGES_ENABLED = 1 << 3,
+    CODE_INSIGHTS_ENABLED = 1 << 4,
+}
```
Very nice!

sqs commented:
Thanks! Seeing how big and annoying these upsells are, I think we should remove them entirely even if Cody is enabled. It just feels tacky to have a big non-dismissible ad (that looks like an ad) on something you paid for. I will finish up https://github.com/sourcegraph/sourcegraph/pull/63430 and then will come back to this PR.

sqs commented:
@fkling OK, https://github.com/sourcegraph/sourcegraph/pull/63430 is ready. The PR description has the desired behavior. I also added tests for GlobalNavbar that you can copy if you want. You should also remove the CodyUpsell component altogether; it just feels like an ad.

fkling commented:
@camdencheek, @vovakulikov I made quite a few changes (rabbit hole) to this and would love another review. I could have probably done some of the refactoring in a separate PR but it's too late now.

tl;dr:

- Main nav "modes" are now all CSS
- Main nav is a normal navigation menu
- Submenu's appear on hover
- Submenu "parent" is now clickable for faster navigation

The biggest change now is that I refactored the implementation of the main navigation. I've noticed before that we don't show the dropdown/fly-out menus anymore on the search homepage, which was probably an oversight. I wanted to fix that and also incorporate my suggestions from [this Slack thread](https://sourcegraph.slack.com/archives/C05EA9KQUTA/p1719480896995599?thread_ts=1719477635.382499&cid=C05EA9KQUTA).
This resulted in changing the implementation in such a way that the both the sidebar "mode" and the top nav "mode" now use the same component/structure. The style change is done with CSS only.
I used https://www.w3.org/WAI/tutorials/menus/flyout/ and https://web.dev/articles/website-navigation as guidance. From an a11y perspective the biggest difference is that this menu is now a plain navigation menu, not an application menu that can be navigated by keyboard, which I think is more correct.

I added test to ensure that the cody navigation entries work as explained in https://github.com/sourcegraph/sourcegraph/pull/63430


https://github.com/sourcegraph/sourcegraph/assets/179026/951dc2cb-c764-4a60-b87f-97e4b714f60e

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -14,57 +14,114 @@
+        <!-- Additional wrapper needed to handle sidebar navigation mode -->
+        <div
+            class="content"
+            use:onClickOutside={{ enabled: sidebarNavigationOpen }}
```
Is there no way to conditionally apply an action?

fkling replied:
No 😞

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -14,57 +14,114 @@
+
     $: withCustomContent = $navigationModeStore === NavigationMode.WithCustomContent
-</script>
+    $: sidebarMode = withCustomContent || $isViewportMediumDown
```
Can't this just be implemented directly in CSS? Something like:

```scss
@media (max-width: 768px), 
.withCustomContent {
    ...
}
```

Cool use of a readable though 🙂

fkling replied:
Unfortunately you can combine media queries with selectors. I'd have to duplicate all the CSS for the sidebar.

camdencheek replied:
Huh, that's annoying. I'd always thought that media selectors were "just another CSS selector"

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -14,57 +14,114 @@
+    <nav aria-label="Main" class:as-sidebar={sidebarMode} class:open={sidebarNavigationOpen}>
+        <!-- Additional wrapper needed to handle sidebar navigation mode -->
+        <div
+            class="content"
```
IIRC, we have a global style with that applies stuff to `.content`. Did we get rid of that?

fkling replied:
Yep!

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -143,21 +198,216 @@
+
+        .sub-navigation {
+            :global(a) {
+                padding-left: 3.7rem;
```
This is a very specific number 😄 Is it specific for a reason?

fkling replied:
I just copied all that from the deleted `GlobalSidebarNavigation.svelte` component. You have to ask @vovakulikov :D

camdencheek commented on client/web-sveltekit/src/lib/stores.ts:
```diff
@@ -81,3 +82,17 @@ export function createLocalWritable<T>(localStorageKey: string, defaultValue: T)
+ * Media query store that updates when the media query matches.
+ */
+export function mediaQuery(query: string): Readable<boolean> {
+    const mediaQuery = window.matchMedia(query)
```
TIL about `window.matchMedia`
